### PR TITLE
Boto/Boto3 metadata customization

### PR DESCRIPTION
**Pull request by roxel**
I added `update_headers` and `update_parameters` methods in Boto/Boto3 storages. These methods allow to implement metadata customization for specific files in inheriting classes without rewriting save method. The way this customization is performed can depend on project needs, but in a most basic scenario you can just use an if-else statement checking saved file name and contents.

### CODE REVIEW

File: storages/backends/s3boto.py [hunk 1]
```diff
@@ -494,3 +495,9 @@ def get_available_name(self, name, max_length=None):
             name = self._clean_name(name)
             return name
         return super(S3BotoStorage, self).get_available_name(name, max_length)
+
+    def update_headers(self, headers, name=None, content=None):
+        """
+        Method that can be used to adjust file parameters.
+        """
+        return headers
```
Comment (mikekeda) [inline storages/backends/s3boto.py]: Probably we can move headers update functionality from _save() into this method

```
    def update_headers(self, name=None, content=None, content_type=None):
        """
        Method that can be used to adjust file parameters.
        """
        headers = self.headers.copy()

        # setting the content_type in the key object is not enough.
        headers.update({'Content-Type': content_type})
 
        if self.gzip and content_type in self.gzip_content_types:
            headers.update({'Content-Encoding': 'gzip'})
        elif encoding:
            # If the content already has a particular encoding, set it
            headers.update({'Content-Encoding': encoding})

        return headers
```

File: storages/backends/s3boto3.py [hunk 2]
```diff
@@ -580,3 +581,9 @@ def get_available_name(self, name, max_length=None):
             name = self._clean_name(name)
             return name
         return super(S3Boto3Storage, self).get_available_name(name, max_length)
+
+    def update_parameters(self, parameters, obj=None, content=None):
+        """
+        Method that can be used to adjust file parameters.
+        """
+        return parameters
```
Comment (mikekeda) [inline storages/backends/s3boto3.py]: Probably we can move parameters update functionality from _save() into this method

```
    def update_parameters(self, obj=None, content=None, content_type=None):
         """
         Method that can be used to adjust file parameters.
         """
        parameters = self.object_parameters.copy()

         # setting the content_type in the key object is not enough.
         parameters.update({'ContentType': content_type})
 
         if self.gzip and content_type in self.gzip_content_types:
             parameters.update({'ContentEncoding': 'gzip'})
         elif encoding:
             # If the content already has a particular encoding, set it
             parameters.update({'ContentEncoding': encoding})

         return parameters
```

File: storages/backends/s3boto.py [hunk 3]
```diff
@@ -494,3 +495,9 @@ def get_available_name(self, name, max_length=None):
             name = self._clean_name(name)
             return name
         return super(S3BotoStorage, self).get_available_name(name, max_length)
+
+    def update_headers(self, headers, name=None, content=None):
+        """
+        Method that can be used to adjust file parameters.
```
Comment (mikekeda) [inline storages/backends/s3boto.py]: should it be "Method that can be used to adjust file **headers**."?

Comment (roxel) [reply storages/backends/s3boto.py]: yeah, right, I'll change that

File: storages/backends/s3boto.py (hunk 1 above)
Comment (roxel) [reply storages/backends/s3boto.py]: yes. this approach would make `save` cleaner. I'm conviced.

**Comment by roxel** [comment]
@mikekeda what do you think about current implementation?

**Comment by roxel** [comment]
@jschneier Is it something we can merge to master and release soon? Or do you think some updates/fixes are needed here?

**Comment by jschneier** [comment]
The general idea was right and I implemented something similar in #828.

**Pull request closed without merging**